Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+
+The existing implementation is using `IsSupported` to determine whether an existing
+namespace is supported on a given platform.
+
+The granularity is kept at a large scale to have large blocks of functionality controlled
+by a single `IsSupported`.
+
+#### Intrinsic granularity
+
+In general individual intrinsic will be chosen to be fine grained.  These will generally correspond to a single assembly instruction.
+
+#### Intrinsics & Crossgen
+
+For any intrinsic which may not be supported on all variants of a platform.  Crossgen
+Method compilation must be trapped, so that the JIT is forced to generate optimal platform dependent 
+code at runtime.
+
+#### API review
+
+Intrinsics will extend the API of CoreCLR.  They will need to follow standard API review practices.
+
+### Choice of Arm64 naming conventions
+
+#### Namespaces
+
+`ARM32` and `ARM64` will follow similar namespace conventions.
+
++ `System.Runtime.Intrinsics` is used for type definitions useful across multiple platforms
++ `System.Runtime.Intrinsics.Arm` is used for enumerations shared across `ARM32` and `ARM64` platforms
++ `System.Runtime.Intrinsics.Arm.Base` is recommended for functionality included in all `ARM32` and `ARM64` platforms
+  + Do we need this or could we assume `ARMv8` is minimum supported platform?
+  + Should we assume `ARMv7` or `ARMv5` is minimum supported platform?
++ `System.Runtime.Intrinsics.Arm.ARMv8` is used for `ARMv8` added functionality included in both `ARM32` and `ARM64` platforms 
++ `System.Runtime.Intrinsics.Arm.ARMv8.Arm64` is used for `ARMv8` functionality only included in `ARM64` platform
++ `System.Runtime.Intrinsics.Arm.ARMv8_1....` is used for `ARMv8.1` functionality.  
+  + Use similar sub namespace names for only included in `ARM64` platform ...
++ `System.Runtime.Intrinsics.Arm.XXX` is used for optional features `CRC32`, `AES`, `SHA1` ... 
+  + All extensions which were not required when they were initially introduced
+  + For example, `CRC32` became required with `ARMv8.1`, but since it was introduced as optional in `ARMv8` it would go in its own namespace
+  
+If an extension is only supported on `Arm32`, there is no reason to add a sub namespace.
+
++ `System.Runtime.Intrinsics.Arm.ARMv7` is used for `ARMv7` added functionality not included in `Base`
+
+If an extension is only supported on `Arm64`, there is no reason to add a sub namespace.  Hypothetically:
+
++ `System.Runtime.Intrinsics.Arm.ARMv10` is used for `ARMv10` added functionality (Hypothetically `Arm64` only)
+
+Namespace is chosen based on what could be implemented, not on what is or will be implemented.
+
+#### Names
+
+Intrinsics will be named to describe functionality.  Names will not correspond to specific named assembly instructions.
+
+`IsSupported` will be used in any namespace which may not be implemented on all CoreCLR platforms
+
+### Intrinsic Interface Documentation
+
++ Documentation will be minimal preferring the underlying ARM documentation
```

Comment:
I am very much against it as Arm tends to write hard to read bloated documentation. IMO we should make the effort and translate Arm engineering language into english language here :smile; with size reduction by at least factor of 5.

ARMv8-A profile documentation has almost 7 000 pages. ISA extensions descriptions tend to refer heavily to other parts of the documentation what forces reader to go through several hundreds of pages to get even superficial understanding of the extensions. I have made several attempts to read it and documentation is very hard to follow and understand.

Reply:
It is not practical to rewrite the documentation of every ARM assembly instruction.  It is error prone and risky.  It may not even be permitted by Arm.

While the engineering documents are often incomplete and difficult to understand, the ARM reference manual is evolving.  The instruction descriptions are fairly easily understood.  Of course bleeding edge extensions like SVE are not as easily understood.   However, we do have the luxury that SVE hardware is not yet enabled in commercial hardware AFAIK.  So CoreCLR does not yet need to implement it,